redis/src/cluster_async/mod.rs

```diff
@@ -1303,50 +1303,59 @@ where
         check_existing_conn: bool,
     ) {
         info!("Started refreshing connections to {:?}", addresses);
-        let connections_container = inner.conn_lock.read().await;
-        let cluster_params = &inner.cluster_params;
-        let subscriptions_by_address = &inner.subscriptions_by_address;
-        let glide_connection_optons = &inner.glide_connection_options;
+        let mut tasks = FuturesUnordered::new();
+        let inner = inner.clone();
 
-        stream::iter(addresses.into_iter())
-            .fold(
-                &*connections_container,
-                |connections_container, address| async move {
-                    let node_option = if check_existing_conn {
-                        connections_container.remove_node(&address)
-                    } else {
-                        None
-                    };
+        for address in addresses.into_iter() {
+            let inner = inner.clone();
 
-                    // override subscriptions for this connection
-                    let mut cluster_params = cluster_params.clone();
-                    let subs_guard = subscriptions_by_address.read().await;
-                    cluster_params.pubsub_subscriptions = subs_guard.get(&address).cloned();
-                    drop(subs_guard);
-                    let node = get_or_create_conn(
-                        &address,
-                        node_option,
-                        &cluster_params,
-                        conn_type,
-                        glide_connection_optons.clone(),
-                    )
-                    .await;
-                    match node {
-                        Ok(node) => {
-                            connections_container
-                                .replace_or_add_connection_for_address(address, node);
-                        }
-                        Err(err) => {
-                            warn!(
-                                "Failed to refresh connection for node {}. Error: `{:?}`",
-                                address, err
-                            );
-                        }
-                    }
-                    connections_container
-                },
-            )
-            .await;
+            tasks.push(async move {
+                let connections_container = inner.conn_lock.read().await;
```

Comment:
Making a lock "public" is not a good idea. We should an atomic API and not the lock itself.

For example:

```rust
fn do_something() -> Result<(), Box<dyn Error>>{
  let _lk = self.lock.write()?;
  ...
}

```

if the "something" is complex, we should add an API:

```rust
fn write_lock_and_do<F>(callback: F) -> Result<(), Box<dyn Error>>
    where F: Fn() -> Result<(), Box<dyn Error>> {
    let _lk = self.lock.write()?;
    callback()
}
```

this way we have a full control over the lock and we can avoid misuse of the lock

Reply:
I think that's a good idea, lets do it in a different PR